File: topics/binarytree/src/maxpathsum.cpp
/*
 * from oj.leetcode. given a binary tree, find the maximum path sum. the path may start and end at any node in the tree;
 *
 * test case:
 * 1. root has child,  val may be negetive, positive, zero.
 * 2. root without child, if its val is negative, the result has to be it.
 * */
#include <iostream>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL){}
};

class Solution{
public:
    int maxPathSum(TreeNode *root){
        int res = root->val;    // res must be initialized by root's val! for test case that negative root without child
        maxPath(root, &res);
        return res;
    }

private:
    /*
     * return: max sum of path which ends at *root
     * @param pSum: max sum of any path existing in the tree beneath *root
     * */
    int maxPath(TreeNode *root, int *pSum){
        if(!root)    return 0;

        int leftsum  = maxPath(root->left, pSum);
        int rightsum = maxPath(root->right, pSum);
        *pSum = max(*pSum, root->val + max(leftsum, 0) + max(rightsum, 0));

        int mx = max(leftsum, rightsum);
        return max(root->val + max(mx, 0), 0);
    }
};

File: leetcode/java/maxpathsum.java
/*
 * given a binary tree, find the maximum path sum. the path can start and end at any node in the tree
 * 
 * */
import java.io.*;
import java.util.*;

class TreeNode{
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x){val=x;}
}

public class maxpathsum{
    private int ret=0;
    private int max(int a, int b){
        return a>b ? a : b;
    }
    private int DFS(TreeNode root){
        if(root==null)        return 0;
        int maxl = DFS(root.left);
        int maxr = DFS(root.right);
        ret = max(ret, root.val + max(0, maxl) + max(0, maxr));
        return max(0, root.val + max(0, max(maxl, maxr)));
    }
    public int maxSum(TreeNode root){
        ret = root.val;
        DFS(root);
        return ret;
    }
    public void test_01(){
        
    }
    public static void main(String[] args){
        maxpathsum mps = new maxpathsum();
        mps.test_01();
    }
}
